tc/autotuner/cuda/autotuner.cc

```diff
@@ -0,0 +1,174 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "tc/autotuner/autotuner.h"
+
+#include <atomic>
+#include <chrono>
+#include <numeric>
+#include <thread>
+
+#include <cuda_runtime_api.h>
+#include <glog/stl_logging.h>
+
+#include "tc/autotuner/utils.h"
+#include "tc/core/compiler.h"
+#include "tc/core/cuda/cuda.h"
+#include "tc/core/cuda/cuda_mapping_options_cpp_printer.h"
+#include "tc/core/cuda/cuda_tc_executor.h"
+#include "tc/core/flags.h"
+#include "tc/core/polyhedral/cuda/mapping_types.h"
+#include "tc/core/scope_guard.h"
+#include "tc/core/tensor.h"
+#include "tc/core/utils/math.h"
+#include "tc/lang/canonicalize.h"
+
+namespace tc {
+namespace autotune {
+namespace detail {
+
+using CudaTuningHarness = class TuningHarness<tc::CudaBackend>;
+
+template <>
+typename CudaBackend::MappingOptionsType makeOptions<CudaBackend>(
+    const typename CudaBackend::MappingOptionsType& baseMapping,
+    const CandidateConfiguration& c) {
+  auto options = baseMapping;
+  c.configuration.applyToCudaMappingOptions(options);
```

Comment:
This looks like TuningConfiguration must be made aware of different backend types... Not sure this is the as clean as other separations you've introduced so far.

Reply:
Yes and I have been punting on rewriting this part too. I'd much rather land what we have now and collect the last crumbs as I finally do the functional CPU path.